tamr_client/datasets/dataframe.py

```diff
@@ -0,0 +1,38 @@
+"""
+Convenient functionality for interacting with pandas DataFrames.
+"""
+
+import json
+
+import pandas as pd
+
+import tamr_client as tc
+from tamr_client.datasets.dataset import Dataset
+from tamr_client.datasets.record import PrimaryKeyNotFound
+from tamr_client.types import JsonDict
+
+
+def upsert(
+    session: tc.Session, dataset: Dataset, df: pd.DataFrame, *, primary_key_name: str
```

Comment:
@skalish to support #333 , we could make `primary_key_name` optional and if its omitted use the dataframe's index

Reply:
I'm concerned that using the dataframe index as primary key by default is prone to accidental misuse.  Implementation is pretty simple, but I think at most I could see this as an option that needs to be explicitly set, e.g.
```python
df = pd.DataFrame({"val": [11, 12]})
tc.dataframe.upsert(s, dataset, df, index_as_primary_key=True)
```
where `primary_key` and `index_as_primary_key` are both optional kwargs (default=`None`) with some logic to enforce that exactly one must be passed.

This is definitely a discussion that could benefit from more field input.

Reply:
I will also bring whatever functionality we settle on here into #339 so the `tuc` DataFrame support will operate in the same way.

Reply:
I don't like including both a `primary_key_name` param and a `index_as_primary_key` param because it allows non-sensical calls like `upsert(..., primary_key_name="non-index column", index_as_primary_key=True)`.

Actually, for upsert, we already have the primary key for the dataset that we are updating via `dataset.key_attribute_names`, so we could make it optional and if omitted, look for that column name.

So this gracefully handles all 3 cases:

```python
# 1. Implicitly use the datasets key attribute name as a primary key
tc.dataset(upsert(s, dataset, df)

# 2. Explicitly name a non-index column as a primary key
tc.dataframe.upsert(s, dataset, df, primary_key_name="non-index pk")

# 3. Opt-in to using the dataframe index as the primary key
tc.dataframe.upsert(s, dataset, df, primary_key_name=df.index.name) 
```

Reply:
I recall there being some concern that the convenience of case 1 is probably outweighed by the potential for user error.  That is, if there is a column name in `df` that matches the dataset primary key, but does not actually correspond to it, this would result in overwriting specific records unintentionally.

Reply:
What actually happens if you directly call the API to upsert to a dataset that has an attribute with the same name as the primary key, but uses a separate attribute as the primary key for the upsert? does the upsert-specified PK get used to locate the record but then the PK gets overwritten by the attribute with the same name?

I imagine its pretty common for FEs to name their pks the same thing across creation/upserts, but we could verify this by talking to them.

Reply:
I just tested this behavior. The API will not overwrite the PK, even though I specified the PK field (with a different value) in the record. However, there was no error. The record with the specified recordId was overwritten everywhere I specified except the new PK I had specified.
To put this another way, it seems like every field of the record EXCEPT the PK is eligible to be overwritten.

Reply:
I think @pcattori's suggested way of handling `primary_key_name` is good.  One consequence that we need to handle is that a DataFrame's index name can be the same as the name of a column, and so this leads to a possible ambiguity.  We could handle this by checking if `primary_key_name` matches both `df.index.name` and a column name in `df.columns` and raise a `ValueError` if there is such a collision.  Incoming commit soon.